Project environment

bishax opened the pull request:
Closes #2

jaklinger commented on metaflow_custom/utils.py:
```diff
@@ -0,0 +1,71 @@
+
+def walk(root: os.PathLike, addl_suffixes: List[str]) -> Iterator[Tuple[str, str]]:
+    """Recursive walk through `root`, yield paths with suffixes in `addl_suffixes`."""
+    prefixlen = len(str(root)) + 1
+    for path, _, files in os.walk(root):
+        for fname in files:
+            if any(fname.endswith(suffix) for suffix in addl_suffixes):
+                p = os.path.join(path, fname)
+                yield p, p[prefixlen:]
```
Suggest to unindent by one, either by:

```python
    prefixlen = len(str(root)) + 1
    suffixes = tuple(addl_suffixes)  # for use in str.endswith
    for path, _, files in os.walk(root):
        for fname in filter(lambda fname: fname.endswith(suffixes), files):
            p = os.path.join(path, fname)
            yield p, p[prefixlen:]
```

or

```python
    prefixlen = len(str(root)) + 1
    for path, _, files in os.walk(root):
        for fname in files:
            if not any(fname.endswith(suffix) for suffix in addl_suffixes):
                continue
            p = os.path.join(path, fname)
            yield p, p[prefixlen:]
```

jaklinger commented on metaflow_custom/utils.py:
```diff
@@ -0,0 +1,71 @@
+import os
```
Could do with some unit tests here

jaklinger reviewed (commented):
A nifty feature you've found here, nice. I've left some comments regarding 

* using `BatchDecorator` as the base class - I can't think of a case where that won't be useful (am I missing something there?)
* small refactor to unindent one function either by reversing logic or using `filter`
* suggestion to add unit tests to `utils` as it underpins the whole mechanism

bishax commented on metaflow_custom/plugins/install_step_decorator.py:
```diff
@@ -0,0 +1,57 @@
+from metaflow_custom.utils import install_setup_py, running_in_local_env, up_to_setup_py
+
+
+class InstallProjectStepDecorator(StepDecorator):
```
It will also be used in conjunction with `@conda` (and potentially any other future decorators that are released, e.g. K8S).

I might be misunderstanding what you mean by boilerplate and absent-minded errors but the beauty is that there isn't any boilerplate...

* `metaflow_custom.config.metaflow_config.DEFAULT_ENVIRONMENT` uses `ProjectEnvironment` by default.
* `ProjectEnvironment.decospecs` automatically adds `@install_project` to every step in the flow.

bishax replied:
So with `metaflow_custom` installed then this decorator gets added to every step by default.

(It can be over-ridden either by setting `DEFAULT_ENVIRONMENT` env var or passing `--environment local`)

jaklinger replied:
> So with metaflow_custom installed then this decorator gets added to every step by default.

ah gotcha!

bishax commented:
> A nifty feature you've found here, nice. I've left some comments regarding
> 
>     * using `BatchDecorator` as the base class - I can't think of a case where that won't be useful (am I missing something there?)
> 
>     * small refactor to unindent one function either by reversing logic or using `filter`
> 
>     * suggestion to add unit tests to `utils` as it underpins the whole mechanism

@jaklinger thanks for the review.
Any thoughts on:
- The definition of a project?
  - It might be best to have a more open interface where `METAFLOW_PROJECT_FILES = ["setup.py"]` is overrideable like other metaflow variables, and these define the project base. This means if we (or any user) switch to another tool such as poetry which uses `pyproject.toml` then this can be a user change.
- Anything from an `ojd_daps` perspective? i.e. if we had this extension mechanism at the start of the project, would it serve your needs?
  -  The main one remaining that I can think of is the ability to install flow specific pip requirements. This can be implemented as a step and flow decorator similar to `conda` and `conda_base`, which is orthogonal to this PR.

jaklinger commented:
One thing that is missing in this PR are some recommended instructions for developmental use of this feature, so that we can "see it in action" - would you be able to write down your recipe for doing that?

jaklinger commented:
> Any thoughts on:
> 
> * The definition of a project?
>   
>   * It might be best to have a more open interface where `METAFLOW_PROJECT_FILES = ["setup.py"]` is overrideable like other metaflow variables, and these define the project base. This means if we (or any user) switch to another tool such as poetry which uses `pyproject.toml` then this can be a user change.

This did cross my mind, one option would be to put an empty file `.PROJECT_HEAD` or similar at the head of the project, which would then allow for submodules

> * Anything from an `ojd_daps` perspective? i.e. if we had this extension mechanism at the start of the project, would it serve your needs?

Not that comes to mind right now, the functionality should serve the purpose as is (i.e. to allow for more intuitive importing syntax). The only additional feature request is in your next point :)

>   * The main one remaining that I can think of is the ability to install flow specific pip requirements. This can be implemented as a step and flow decorator similar to `conda` and `conda_base`, which is orthogonal to this PR.

Sweet - looking forward to that! One request on that feature is where there are multiple `requirements_*.txt` to be installed per flow (where two flows in the same dir may have both common and divergent requirements)

> * Better documenting how `metaflow_custom` works in general

As per my previous comment: for this PR how we can double check that this works as expected in development.

Also perhaps in the README just the `installation` section for this repo, plus what the example snippet looks like after being fixed by this package.

bishax commented:
> One request on that feature is where there are multiple requirements_*.txt to be installed per flow (where two flows in the same dir may have both common and divergent requirements)

Did you know that `-r common_requirements.txt` is a valid line in a requirements.txt format and will install requirements from `common_requirements.txt` (name not a special case), so the above feature shouldn't be needed?

jaklinger commented:
I mean the case when there are multiple `requirement_*.txt` e.g.:

```
requirements_common.txt
requirements_a.txt
requirements_b.txt
```

and then being able to do:

```python
@conda("requirements_common.txt", "requirements_a.txt")
def FlowA(FlowSpec):
    ...

@conda("requirements_common.txt", "requirements_b.txt")
def FlowB(FlowSpec):
    ...
```

bishax commented:
I got what you mean, what I was getting at was that that use-case is captured by doing the following:

```
#file: requirements_common.txt

luigi
toolz
yada-yada

#file: requirements_a.txt

pytorch
-r requirements_common.txt

#file: requirements_b.txt

tensorflow
-r requirements_common.txt
```

```python
@tbd_decorator_name("requirements_a.txt")
def FlowA(FlowSpec):
    ...

@tbd_decorator_name("requirements_b.txt")
def FlowB(FlowSpec):
    ...
```

Which means:
- We have to write less code
- The requirements files are more explicit and less prone to user-error (forgetting to include the commons in a flow file)

bishax commented:
Source: https://pip.pypa.io/en/stable/cli/pip_install/#install-requirement

bishax commented:
@jaklinger addressed changes (refactor + tests for utils + clearer docs).

Hopefully the docs are now clear enough that just installing `metaflow-custom` is enough to make a flow work with batch/conda?

jaklinger commented:
Looks good - what are your thoughts on the suggestion

> This did cross my mind, one option would be to put an empty file .PROJECT_HEAD or similar at the head of the project, which would then allow for submodules

bishax commented:
> Looks good - what are your thoughts on the suggestion
> 
> > This did cross my mind, one option would be to put an empty file .PROJECT_HEAD or similar at the head of the project, which would then allow for submodules

@jaklinger what do you mean by allowing for submodules?

`.PROJECT_HEAD` works for adding files to the job package but doesn't necessarily work for installation - what if there is a `.PROJECT_HEAD` but no `pip install`-able file?

How about defining `METAFLOW_PROJECT_FILES=["setup.py", "pyproject.toml"]` as a config variable (like other metaflow variables)? By default, it only has pip-installable files (so the directory packaged and installed won't fail on `pip install`) but it is also open to change by user-code.

jaklinger commented:
> @jaklinger what do you mean by allowing for submodules?
> 
> `.PROJECT_HEAD` works for adding files to the job package but doesn't necessarily work for installation - what if there is a `.PROJECT_HEAD` but no `pip install`-able file?

I'm thinking that many data science repos won't necessarily be installable, but then in which case this method won't work (or at least it presents a (low but I suspect significant) technical barrier to entry) - but you're right in that there isn't really much of a workaround for that

> How about defining `METAFLOW_PROJECT_FILES=["setup.py", "pyproject.toml"]` as a config variable (like other metaflow variables)? By default, it only has pip-installable files (so the directory packaged and installed won't fail on `pip install`) but it is also open to change by user-code.

I think that's a good idea

bishax commented:
> I'm thinking that many data science repos won't necessarily be installable
`ds-cookiecutter` is designed to be pip-installable without user setup so we shouldn't hit that technical barrier to entry :crossed_fingers:

jaklinger commented:
ok cool, the tests run as well with the recent fix in #6 so i think we're good to go

bishax commented:
@jaklinger I implemented the change proposed above.

Diff here (also rebased onto the changes from the setup PR): https://github.com/nestauk/metaflow_custom/pull/7/files/6c242c506ac450e9e496c013202e5c6aed1957fb..2135c099b813442347630929ac3967f558a8b436

Hopefully changes shouldn't be controversial.

jaklinger commented:
LGTM!
